plugins/user-settings/src/components/useUserProfileInfo.ts

```diff
@@ -14,13 +14,36 @@
  * limitations under the License.
  */
 
-import { useApi, identityApiRef } from '@backstage/core-plugin-api';
+import {
+  useApi,
+  identityApiRef,
+  BackstageUserIdentity,
+  ProfileInfo,
+} from '@backstage/core-plugin-api';
+import { useEffect, useState } from 'react';
 
 export const useUserProfile = () => {
   const identityApi = useApi(identityApiRef);
-  const userId = identityApi.getUserId();
-  const profile = identityApi.getProfile();
-  const displayName = profile.displayName ?? userId;
+
+  const [displayName, setDisplayName] = useState<
+    string | BackstageUserIdentity
+  >('');
+  const [profile, setProfile] = useState<ProfileInfo>({});
+
+  const getUserProfile = async () => {
+    const backstageIdentity = await identityApi.getBackstageIdentity();
+    const profileInfo = await identityApi.getProfileInfo();
+    const name = profileInfo.displayName ?? backstageIdentity;
+
+    setDisplayName(name);
+    setProfile(profileInfo);
+  };
+
+  useEffect(() => {
+    if (!displayName) {
+      getUserProfile();
+    }
+  });
```

Comment:
Hm this all turns out a bit awkward.

May I suggest using useAsync instead?

```ts
const alertApi = useApi(alertApiRef);

const { value, loading, error } = useAsync(async () => {
 return {
   profile: await identityApi.getProfileInfo();
   identity: await identityApi.getBackstageIdentity();
 };
}, []);

useEffect(() => {
  if (error) {
    alertApi.post({ message: `Failed to load user identity: ${error}` });
  }
}, [error]);
```

Now the question is what to do with the `value`. We could choose to return some empty placeholder values as you do now,

```ts
if (loading) {
  return ...
}
```

And maybe that's a good first stab at it. However, I would like to _also_ return the `loading` flag.

```ts
return {
  profile: loading ? {} : value.profile,
  displayName: loading ? '' : value.profile?.displayName ?? value.identity.userEntityRef,
  loading,
};
```

that way, callers can choose to show a loading progress bar or similar while waiting for data to arrive.

You'll have to run `yarn build:api-reports` and commit the results after, then, because this amends the contract a bit.

Reply:
Thanks for the pointers! That looks a lot better: I've applied the changes almost verbatim. I've used non-null assertion operators on the `value`s, as I think think they are always non-null when loading is null, right?